### ISSUE TITLE
Add ability to let ignore index be outside of number of classes

### ISSUE THREAD

**Issue by NielsRogge** [Lightning-AI/metrics#935]
## 🚀 Feature

Hi,

So I wanted to use the [Jaccard index](https://torchmetrics.readthedocs.io/en/latest/references/modules.html#jaccardindex) for evaluating my semantic segmentation model. However, regarding the `ignore_index` argument, in the docs it is said that:

> [](https://torchmetrics.readthedocs.io/en/latest/references/modules.html#torchmetrics.JaccardIndex.params.ignore_index)ignore_index ([Optional](https://docs.python.org/3/library/typing.html#typing.Optional)[[int](https://docs.python.org/3/library/functions.html#int)]) – optional int specifying a target class to ignore. If given, this class index does not contribute to the returned score, regardless of reduction method. Has no effect if given an int that is not in the range [0, num_classes-1]. By default, no index is ignored, and all classes are used.

=> it says "Has no effect if given an int that is not in the range [0, num_classes-1]"

However, I'd like to specify the `ignore_index` to be 255 (which is outside of the number of classes - which is 150 as I'm training on [ADE20k-150](http://sceneparsing.csail.mit.edu/)), as targets are created by reducing all labels by 1, and replacing the background class (0) by 255. This is how common semantic segmentation libraries such as [mmsegmentation](https://github.com/open-mmlab/mmsegmentation) do it as well.

To give some more info: the ADE20k dataset consists of annotation maps that are labeled with numbers between 0 and 150, with 0 being the "background" class. Usually, models are only trained on labels 1 - 150, and not on the background class. Hence, one replaces 0 by 255 (which is the ignore index of the PyTorch's model's loss function), and reduces all labels by 1, such that the PyTorch tensors of the annotation maps contain values between 0 and 149 and 255 for everywhere there's background.

**Comment by NielsRogge** [comment]
Awesome, thanks @SkafteNicki

**Comment by vilim** [comment]
This issue persists for the Dice metric which still has a check 
https://github.com/Lightning-AI/torchmetrics/blob/d1c5dd9811c38c7148c8cf83096bf3124db53386/src/torchmetrics/functional/classification/dice.py#L188C1-L188C106
